src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,543 @@
+            JSON::JArray(d) => encode_array(d),
+            JSON::JDouble(d) => encode_double(d),
+            JSON::JI64(d) => encode_i64(d),
+        };
+        let len = encode_data.len();
+        data.append(&mut encode_data);
+        len
+    }
+
+    fn decode(code_type: u8, data: &[u8]) -> Result<JSON> {
+        match code_type {
+            TYPE_CODE_LITERAL => decode_literal(data),
+            TYPE_CODE_STRING => decode_str(data),
+            TYPE_CODE_OBJECT => decode_obj(data),
+            TYPE_CODE_ARRAY => decode_array(data),
+            TYPE_CODE_DOUBLE => decode_double(data),
+            TYPE_CODE_I64 => decode_i64(data),
+            _ => Err(invalid_type!("unsupported type {:?}", code_type)),
+        }
+    }
+
+    fn get_type_code(&self) -> u8 {
+        match *self {
+            JSON::JLiteral(_) => TYPE_CODE_LITERAL,
+            JSON::JString(_) => TYPE_CODE_STRING,
+            JSON::JObject(_) => TYPE_CODE_OBJECT,
+            JSON::JArray(_) => TYPE_CODE_ARRAY,
+            JSON::JDouble(_) => TYPE_CODE_DOUBLE,
+            JSON::JI64(_) => TYPE_CODE_I64,
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        if let JSON::JLiteral(ref d) = *self {
+            match *d {
+                JSON_LITERAL_NIL => return String::from("null"),
+                JSON_LITERAL_TRUE => return String::from("true"),
+                JSON_LITERAL_FALSE => return String::from("false"),
+                _ => return String::from("illegal"),
+            };
+        }
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+impl Serialize for JSON {
+    fn serialize<S>(&self, serializer: S) -> SResult<S::Ok, S::Error>
+        where S: Serializer
+    {
+        match *self {
+            JSON::JLiteral(l) => {
+                match l {
+                    JSON_LITERAL_NIL => serializer.serialize_none(),
+                    JSON_LITERAL_TRUE => serializer.serialize_bool(true),
+                    _ => serializer.serialize_bool(false),
+                }
+            }
+            JSON::JString(ref s) => serializer.serialize_str(s),
+            JSON::JObject(ref obj) => {
+                let mut map = try!(serializer.serialize_map(Some(obj.len())));
+                for (k, v) in obj {
+                    try!(map.serialize_entry(k, v));
+                }
+                map.end()
+            }
+            JSON::JArray(ref array) => {
+                let mut tup = try!(serializer.serialize_tuple(array.len()));
+                for item in array {
+                    try!(tup.serialize_element(item));
+                }
+                tup.end()
+            }
+            JSON::JDouble(d) => serializer.serialize_f64(d),
+            JSON::JI64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+fn decode_literal(mut data: &[u8]) -> Result<JSON> {
+    let literal = take_prefix!(data, LENGTH_LITERAL);
+    Ok(JSON::JLiteral(literal[0]))
+}
+
+fn encode_str(data: String) -> Vec<u8> {
```

Comment:
Better pass `buf: &mut Vec<u8>` or `mut buf: Write` as second parameter,
to avoid Vec allocation